### PR TITLE
Aerodynamic forces correction enhancements

### PR DESCRIPTION
**Pull request by ngoiz**
This PR tranfsorms the aerodynamic forces correction functions to generators to be used in the coupled solvers. This is for cases where airfoil polars are defined (see updated Case Files documentation). 

In this PR we include:
 * Correct forces: transform functions to generators
 * Bug fixes in polar corrections and full documentation
 * Update to the `aeroforcescalculator` post-processor to output corrected forces and to output aerodynamic forces and moments. The latter are now written on a separate file `moments_`
 * Changes to the aero data structure such that the aero forces expressed in the beam degrees of freedom are included in the aero time step

### CODE REVIEW

File: sharpy/aero/utils/mapping.py [hunk 1]
```diff
@@ -75,3 +75,55 @@ def aero2struct_force_mapping(aero_forces,
...
+                i_n = mapping['i_n']
+                _, n_m, _ = aero_forces[i_surf].shape
+
+                for i_m in range(n_m):
+                    zeta_g = zeta[i_surf][:, i_m, i_n]
+                    total_aero_forces[0:3] += cag.dot(aero_forces[i_surf][0:3, i_m, i_n])
+                    total_aero_forces[3:6] += cag.dot(aero_forces[i_surf][3:6, i_m, i_n])
+                    total_aero_forces[3:6] += cag.dot(algebra.cross3(zeta_g, aero_forces[i_surf][0:3, i_m, i_n]))
+
+    return total_aero_forces
```
Comment (sduess) [inline sharpy/aero/utils/mapping.py:129]:
```suggestion

```
The function is not used anywhere. You probably have just forgotten to remove it.

File: sharpy/utils/datastructures.py [hunk 2]
```diff
@@ -165,13 +165,15 @@ def __init__(self, dimensions, dimensions_star):
...
-        # total forces
+        # total forces - written by AeroForcesCalculator
         self.inertial_total_forces = np.zeros((self.n_surf, 6))
         self.body_total_forces = np.zeros((self.n_surf, 6))
         self.inertial_steady_forces = np.zeros((self.n_surf, 6))
         self.body_steady_forces = np.zeros((self.n_surf, 6))
         self.inertial_unsteady_forces = np.zeros((self.n_surf, 6))
         self.body_unsteady_forces = np.zeros((self.n_surf, 6))
+        self.total_inertial_forces = np.zeros((6,))  # G Frame
+        self.total_body_forces = np.zeros((6,))  # A Frame
```
Comment (sduess) [inline sharpy/utils/datastructures.py]: These forces are not used anywhere as well are others. You committed the necessary changes in commit 8d1021e30e195a79795c6ed8357c1ac286367e0b

**Review by sduess — CHANGES_REQUESTED** [review]
Good job Norberto! I just added some minor suggestions which are all related to "delete stuff we don't use in our code".

File: sharpy/aero/utils/mapping.py (hunk 1 above)
Comment (ngoiz) [reply sharpy/aero/utils/mapping.py]: Indeed! Thanks for spotting it!

**Review by sduess — APPROVED** [review]

**Review by ArturoMS13 — CHANGES_REQUESTED** [review]
Thank you very much for the contribution. I add some minor comments open to discussion.

File: docs/source/content/casefiles.rst [hunk 3]
```diff
@@ -315,6 +315,14 @@ Item by item:
     For more information on how these factors are included in the mapping terms
     see :func:`sharpy.aero.utils.mapping.aero2struct_force_mapping`.
 
+* ``polars`` Group (optional): Use airfoil polars to correct aerodynamic forces.
+
+    This is an optional group to add if correcting the aerodynamic forces using airfoil polars is desired. A polar
+    should be included for each airfoil defined. Each entry consists of a 4-column table. The first column corresponds
+    to the angle of attack (in radians) and then the ``C_L``, ``C_D`` and ``C_M``. If different airfoils are used a
```
[the remark this answers is not in the document]
Comment (ngoiz) [reply docs/source/content/casefiles.rst]: True! 😅

**Comment by ngoiz** [comment]
Thanks both for your comments. I think I have addressed all your points:
* Horrible `aerotstep.aero_forces_beam_dof` is gone and instead it goes into `structtstep.postproc_nodes['aero_steady_forces]` (or unsteady) which is then plotted in Paraview
* Opted to retain `aero_tstep.body_steady_forces` since those are a `6x` vector and for each surface, which is useful in some cases and it doesn't go in nicely onto the `postproc_node` strategy as above. These are defined in the `AeroTimestepInfo` class.
* Adapted the `AeroForcesCalculator` postproc to run online and offline
* Moved the test to the polars tests `/tests/uvlm/static/polars`

I think this covers my changes. @ArturoMS13 once you are happy with the stability axes and your twisting wings I think this should be then good to go

**Comment by ArturoMS13** [comment]
Thank you for the contribution
I have double checked this implementation for wind turbines and it works except from the root regions. I will fix the root soon, but it is not a big deal so this PR can be merged.

File: sharpy/generators/polaraeroforces.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                                                                'moment_from_polar': 'off'}
+
+    These are the steps needed to correct the forces:
+
+        1. The force coming from UVLM is divided into induced drag (parallel to the incoming flow velocity) and lift
+          (the remaining force).
+
+    If ``cd_from_cl == 'on'``.
+        2. The viscous drag and pitching moment is found at the computed lift coefficient and the forces and
+           moments updated
```
Comment (ArturoMS13) [inline sharpy/generators/polaraeroforces.py:32]:
```suggestion
        2. The viscous drag and pitching moment are found at the computed lift coefficient. Then forces and
           moments are updated
```

File: sharpy/generators/polaraeroforces.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+    Else, the angle of attack is computed:
+
+            2. The angle of attack is computed based on that lift force and the angle of zero lift computed from the
+               airfoil polar and taking the potential flow lift curve slope of :math:`2 \pi`
```
Comment (ArturoMS13) [inline sharpy/generators/polaraeroforces.py:37]:
```suggestion
               airfoil polar and assuming the potential flow lift curve slope of :math:`2 \pi`
```

File: sharpy/generators/polaraeroforces.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+        3. The drag force is computed based on the angle of attack and the polars provided by the user
+
+        4. If ``correct_lift == 'on'``, the lift coefficient is also corrected with the polar data. Else only the
+           UVLM results are used.
+
+    The pitching moment is added in a similar manner to the viscous drag. However, if ``moment_from_polar == 'on'``
```
Comment (ArturoMS13) [inline sharpy/generators/polaraeroforces.py:44]:
```suggestion
    The pitching moment is added in a similar manner as the viscous drag. However, if ``moment_from_polar == 'on'``
```

File: sharpy/generators/polaraeroforces.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+    settings_types['cd_from_cl'] = 'bool'
+    settings_default['cd_from_cl'] = False
+    settings_description['cd_from_cl'] = 'Interpolate the C_D for the given C_L, as opposed to getting the C_D from ' \
+                                         'the section AoA.'
+
+    settings_types['moment_from_polar'] = 'bool'
+    settings_default['moment_from_polar'] = False
+    settings_description['moment_from_polar'] = 'If ``correct_lift`` is selected, it will compute the pitching moment ' \
+                                                'simply from poalr derived data, i.e. the polars Cm and the moments' \
```
Comment (ArturoMS13) [inline sharpy/generators/polaraeroforces.py:69]:
```suggestion
                                                'simply from polar derived data, i.e. the polars Cm and the moments' \
```

File: sharpy/generators/polaraeroforces.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+           moments updated
+
+    Else, the angle of attack is computed:
+
+            2. The angle of attack is computed based on that lift force and the angle of zero lift computed from the
+               airfoil polar and taking the potential flow lift curve slope of :math:`2 \pi`
+
+        3. The drag force is computed based on the angle of attack and the polars provided by the user
+
+        4. If ``correct_lift == 'on'``, the lift coefficient is also corrected with the polar data. Else only the
```
Comment (ArturoMS13) [inline sharpy/generators/polaraeroforces.py:41]:
```suggestion
        4. If ``correct_lift == 'on'``, the lift coefficient is also corrected with the polar data. Else, only the
```

**Review by ArturoMS13 — APPROVED** [review]
Just some extra minor changes. Thank you for the contribution

**Pull request merged**